### PR TITLE
feat: add Jina Embeddings integration

### PR DESCRIPTION
**Pull request by JoanFM**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -64,3 +64,4 @@ deepset-haystack
...
 | [unstructured-fileconverter-haystack](integrations/unstructured/fileconverter/) | File converter      | [![PyPI - Version](https://img.shields.io/pypi/v/unstructured-fileconverter-haystack.svg)](https://pypi.org/project/unstructured-fileconverter-haystack) | [![Test / unstructured / fileconverter](https://github.com/deepset-ai/haystack-core-integrations/actions/workflows/unstructured_fileconverter.yml/badge.svg)](https://github.com/deepset-ai/haystack-core-integrations/actions/workflows/unstructured_fileconverter.yml) |
+| [jina-haystack](integrations/jina/)                                             | Embedder            | [![PyPI - Version](https://img.shields.io/pypi/v/jina-haystack.svg)](https://pypi.org/project/jina-haystack)                                         | [![Test / cohere](https://github.com/deepset-ai/haystack-core-integrations/actions/workflows/cohere.yml/badge.svg)](https://github.com/deepset-ai/haystack-core-integrations/actions/workflows/cohere.yml)                                                               |
```
Comment (anakin87) [inline README.md:67]:
```suggestion
| [jina-haystack](integrations/jina/)                                             | Embedder            | [![PyPI - Version](https://img.shields.io/pypi/v/jina-haystack.svg)](https://pypi.org/project/jina-haystack)                                         | [![Test / cohere](https://github.com/deepset-ai/haystack-core-integrations/actions/workflows/jina.yml/badge.svg)](https://github.com/deepset-ai/haystack-core-integrations/actions/workflows/jina.yml)                                                               |
```

File: integrations/jina/README.md [hunk 2]
```diff
@@ -0,0 +1,32 @@
+# jina-haystack
+
+[![PyPI - Version](https://img.shields.io/pypi/v/cohere-haystack.svg)](https://pypi.org/project/jina-haystack)
+[![PyPI - Python Version](https://img.shields.io/pypi/pyversions/cohere-haystack.svg)](https://pypi.org/project/jina-haystack)
```
Comment (anakin87) [inline integrations/jina/README.md:4]:
```suggestion
[![PyPI - Version](https://img.shields.io/pypi/v/jina-haystack.svg)](https://pypi.org/project/jina-haystack)
[![PyPI - Python Version](https://img.shields.io/pypi/pyversions/jina-haystack.svg)](https://pypi.org/project/jina-haystack)
```

File: integrations/jina/pyproject.toml [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+  "Programming Language :: Python :: 3.11",
+  "Programming Language :: Python :: Implementation :: CPython",
+  "Programming Language :: Python :: Implementation :: PyPy",
+]
+dependencies = ["requests", "haystack-ai"]
+
+[project.urls]
+Documentation = "https://github.com/unknown/jina-haystack#readme"
+Issues = "https://github.com/unknown/jina-haystack/issues"
+Source = "https://github.com/unknown/jina-haystack"
```
Comment (anakin87) [inline integrations/jina/pyproject.toml:32]:
```suggestion
Documentation = "https://github.com/deepset-ai/haystack-core-integrations/tree/main/integrations/jina/jina-haystack#readme"
Issues = "https://github.com/deepset-ai/haystack-core-integrations/issues"
Source = "https://github.com/deepset-ai/haystack-core-integrations/tree/main/integrations/jina/jina-haystack"
```

File: integrations/jina/src/jina_haystack/__init__.py [hunk 4]
```diff
@@ -0,0 +1,3 @@
+# SPDX-FileCopyrightText: 2023-present deepset GmbH <info@deepset.ai>
+#
+# SPDX-License-Identifier: Apache-2.0
```
Comment (anakin87) [inline integrations/jina/src/jina_haystack/__init__.py:3]: Please import the components here.
You can take [elasticsearch-haystack](https://github.com/deepset-ai/haystack-core-integrations/blob/f3ec5a015fa2d3ab5ef9a697fb75a23a3b58ae9f/integrations/elasticsearch/src/elasticsearch_haystack/__init__.py) as an example.

File: integrations/jina/src/jina_haystack/document_embedder.py [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+        batch_size: int = 32,
+        progress_bar: bool = True,
+        metadata_fields_to_embed: Optional[List[str]] = None,
+        embedding_separator: str = "\n",
+    ):
+        """
+        Create a JinaDocumentEmbedder component.
+        :param api_key: The Jina API key. It can be explicitly provided or automatically read from the
+            environment variable JINA_API_KEY (recommended).
+        :param model_name: The name of the Jina model to use.
```
Comment (anakin87) [inline integrations/jina/src/jina_haystack/document_embedder.py:51]: You can add a link to Jina docs to help the user choose the right model.

File: integrations/jina/src/jina_haystack/document_embedder.py [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+        Embed a list of texts in batches.
+        """
+
+        all_embeddings = []
+        metadata = {}
+        for i in tqdm(
+            range(0, len(texts_to_embed), batch_size), disable=not self.progress_bar, desc="Calculating embeddings"
+        ):
+            batch = texts_to_embed[i : i + batch_size]
+            response = self._session.post(  # type: ignore
```
Comment (anakin87) [inline integrations/jina/src/jina_haystack/document_embedder.py:141]: what is the problem with mypy here?

File: integrations/jina/src/jina_haystack/text_embedder.py [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+        model_name: str = "jina-embeddings-v2-base-en",
+        prefix: str = "",
+        suffix: str = "",
+    ):
+        """
+        Create an JinaTextEmbedder component.
+
+        :param api_key: The Jina API key. It can be explicitly provided or automatically read from the
+            environment variable JINA_API_KEY (recommended).
+        :param model_name: The name of the Jina model to use.
```
Comment (anakin87) [inline integrations/jina/src/jina_haystack/text_embedder.py:46]: Same as above. Please provide a link to the docs if available...

File: integrations/jina/tests/test_document_embedder.py [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+        assert embedder.model_name == "model"
+        assert embedder.prefix == "prefix"
+        assert embedder.suffix == "suffix"
+        assert embedder.batch_size == 64
+        assert embedder.progress_bar is False
+        assert embedder.metadata_fields_to_embed == ["test_field"]
+        assert embedder.embedding_separator == " | "
+
+    def test_init_fail_wo_api_key(self, monkeypatch):
+        monkeypatch.delenv("OPENAI_API_KEY", raising=False)
```
Comment (anakin87) [inline integrations/jina/tests/test_document_embedder.py:60]:
```suggestion
```

This only makes sense for the OpenAI Embedders.

File: integrations/jina/src/jina_haystack/document_embedder.py [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+                JINA_API_URL, json={"input": batch, "model": self.model_name}
+            ).json()
+            if "data" not in response:
+                raise RuntimeError(response["detail"])
+
+            # Sort resulting embeddings by index
+            sorted_embeddings = sorted(response["data"], key=lambda e: e["index"])  # type: ignore
```
Comment (anakin87) [inline integrations/jina/src/jina_haystack/document_embedder.py:148]: and here?

File: integrations/jina/src/jina_haystack/document_embedder.py [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+            meta_values_to_embed = [
+                str(doc.meta[key])
+                for key in self.metadata_fields_to_embed
+                if key in doc.meta and doc.meta[key] is not None
+            ]
+            text_to_embed = (
+                self.prefix + self.embedding_separator.join([*meta_values_to_embed, doc.content or ""]) + self.suffix
+            )
+
+            text_to_embed = text_to_embed.replace("\n", " ")
```
Comment (anakin87) [inline integrations/jina/src/jina_haystack/document_embedder.py:126]: This is recommended for OpenAI embedding models.
Jina models require the same?

**Review by anakin87 — COMMENTED** [review]
Thanks for this PR!

I left some comments...

File: integrations/jina/src/jina_haystack/document_embedder.py (hunk 6 above)
Comment (JoanFM) [reply integrations/jina/src/jina_haystack/document_embedder.py]: Removed

File: integrations/jina/src/jina_haystack/text_embedder.py (hunk 7 above)
Comment (JoanFM) [reply integrations/jina/src/jina_haystack/text_embedder.py]: Done

File: integrations/jina/tests/test_document_embedder.py (hunk 8 above)
Comment (JoanFM) [reply integrations/jina/tests/test_document_embedder.py]: Changed the key

File: integrations/jina/src/jina_haystack/document_embedder.py (hunk 9 above)
Comment (JoanFM) [reply integrations/jina/src/jina_haystack/document_embedder.py]: Removed

File: integrations/jina/src/jina_haystack/document_embedder.py (hunk 10 above)
Comment (JoanFM) [reply integrations/jina/src/jina_haystack/document_embedder.py]: Removed, thanks

**Comment by JoanFM** [comment]
> Thanks for this PR!
> 
> I left some comments...

hey @anakin87 , thanks for the review. This is ready again

**Review by anakin87 — APPROVED** [review]
Looks good to me!

I merge it and create the package...

**Comment by anakin87** [comment]
@JoanFM thanks for this integration!

It would be great if you could open a simple PR in https://github.com/deepset-ai/haystack-integrations.
This will help to highlight this integration in the [official Integrations page](https://haystack.deepset.ai/integrations).

You can take [this](https://github.com/deepset-ai/haystack-integrations/blob/main/integrations/gradient.md) as an example.

Let me know if you need any help...

**Comment by JoanFM** [comment]
> @JoanFM thanks for this integration!
> 
> It would be great if you could open a simple PR in https://github.com/deepset-ai/haystack-integrations. This will help to highlight this integration in the [official Integrations page](https://haystack.deepset.ai/integrations).
> 
> You can take [this](https://github.com/deepset-ai/haystack-integrations/blob/main/integrations/gradient.md) as an example.
> 
> Let me know if you need any help...

I have opened one.

**Pull request merged**